File: convert.jl
function convert_heading(s::String)::String
    parts = split(s, "\n"; keepempty=true)
    output = Vector{String}()

    for line in parts
        if startswith(line, "# ") push!(output, "\\section{$(SubString(line, 3))}")
        elseif startswith(line, "## ") push!(output, "\\section{$(SubString(line, 4))}")
        elseif startswith(line, "### ") push!(output, "\\section{$(SubString(line, 5))}")
        elseif startswith(line, "#### ") push!(output, "\\subsection{$(SubString(line, 6))}")
        elseif startswith(line, "##### ") push!(output, "\\subsubsection{$(SubString(line, 7))}")
        else push!(output, line)
        end
    end

    join(output, "\n")
end

function convert_listing(s::String)::String
    parts = split(s, "\n"; keepempty=true)
    output = Vector{String}()
    block = false

    for line in parts
        if startswith(line, "```")
            begin
                if block
                    push!(output, "\\end{lstlisting}")
                    block = false
                else
                    push!(output, "\\begin{lstlisting}")
                    block = true
                end
            end
        else push!(output, line)
        end
    end

    join(output, "\n")
end

function convert_bullets(s::String)::String
    parts = split(s, "\n"; keepempty=true)
    output = Vector{String}()
    block = false

    for line in parts
        if match(r"^\s*\* ", line) !== nothing
            begin
                r = findfirst("*", line)
                if !block push!(output, "\\begin{itemize}\n"); block = true end
                push!(output, "\\item $(SubString(line, r[1]+2))")
            end
        elseif block && isempty(line) push!(output, "\\end{itemize}\n"); block = false
        else push!(output, line)
        end
    end

    join(output, "\n")
end

function convert_link(s::String)::String
    out = s
    m = match(r"\[(.+)\]\((.+)\)", out)

    while m !== nothing
        if m.offset > 1
            rep = "$(SubString(out, 1, m.offset - 1))\\href{$(m[2])}{$(m[1])}$(SubString(out, m.offsets[2]+length(m[2])+1))"
            out = rep
        else
            rep = "\\href{$(m[2])}{$(m[1])}$(SubString(out, m.offsets[2]+length(m[2])+1))"
            out = rep
        end
        m = match(r"\[(.+)\]\((.+)\)", out)
    end

    out
end

function convert_style(line::String, delim::String, cmd::String)::String
    out = ""
    r = findfirst(delim, line)
    if r === nothing return line end

    start = r[1] > 1 ? 1 : r[1] + length(delim)
    r1 = r[1] > 1 ? r[1] - 1 : r[1]
    if r1 < start r1 = start end
    r2 = length(r) > 1 ? r[2] + 1 : r[1] + length(delim)

    e = findnext(delim, line, r2)
    if e === nothing
        out = "$(SubString(line, start, r1)) $(SubString(line, r2))"
    else
        e2 = length(e) > 1 ? e[2] + 1 : e[1] + length(delim)
        out = "$(SubString(line, start, r1))$cmd{$(SubString(line, r2, e[1]-1))}$(SubString(line, e2))"
    end

    return convert_style(out, delim, cmd)
end

function convert_style(line::String)::String
    out = convert_style(line, "**", "\\textbf")
    out = convert_style(out, "__", "\\textbf")
    out = convert_style(out, "*", "\\textit")
    out = convert_style(out, "_", "\\textit")
    out = convert_style(out, "`", "\\texttt")
    out
end

function convert(s::String)::String
    processed = convert_heading(s)
    processed = convert_listing(processed)
    processed = convert_bullets(processed)
    processed = convert_link(processed)

    return convert_style(processed)
end

File: src/output/convert.cpp
//
// Created by Rakesh on 22/09/2025.
//

#include "output.hpp"
#include "../util/split.hpp"

#include <regex>
#include <boost/algorithm/string/classification.hpp>
#include <boost/algorithm/string/replace.hpp>
#include <boost/algorithm/string/split.hpp>

using std::operator ""sv;

namespace
{
  namespace pconvert
  {
    std::string heading( std::string_view data )
    {
      if ( !data.contains( '#' ) ) return std::string{ data };

      auto result = std::string{};
      result.reserve( data.size() );
      auto parts = std::vector<std::string>{};
      parts.reserve( 8 );
      boost::algorithm::split( parts, data, boost::is_any_of( "\r\n" ) );
      auto first = true;

      for ( const auto& line : parts )
      {
        if ( !first ) result.append( "\n" );
        if ( line.empty() ) result.append( "\n" );
        else if ( line.starts_with( "# " ) ) result.append( R"(\section{)"sv ).append( line.substr( 2 ) ).append( "}\n" );
        else if ( line.starts_with( "## " ) ) result.append( R"(\section{)"sv ).append( line.substr( 3 ) ).append( "}\n" );
        else if ( line.starts_with( "### " ) ) result.append( R"(\section{)"sv ).append( line.substr( 4 ) ).append( "}\n" );
        else if ( line.starts_with( "#### " ) ) result.append( R"(\subsection{)"sv ).append( line.substr( 5 ) ).append( "}\n" );
        else if ( line.starts_with( "##### " ) ) result.append( R"(\subsubsection{)"sv ).append( line.substr( 6 ) ).append( "}\n" );
        else result.append( line );
        first = false;
      }

      return result;
    }

    std::string listing( std::string_view data )
    {
      if ( !data.contains( "```" ) ) return std::string{ data };

      auto block = false;
      auto result = std::string{};
      result.reserve( data.size() );
      auto parts = std::vector<std::string>{};
      parts.reserve( 8 );
      boost::algorithm::split( parts, data, boost::is_any_of( "\r\n" ) );
      auto first = true;

      for ( const auto& line : parts )
      {
        if ( !first ) result.append( "\n" );

        if ( line.starts_with( "```" ) )
        {
          if ( block )
          {
            result.append( R"(\end{lstlisting}
)"sv );
            block = false;
          }
          else
          {
            result.append( R"(\begin{lstlisting}
)"sv );
            block = true;
          }
        }
        else result.append( line );
        first = false;
      }

      return result;
    }

    std::string bullets( const std::string& data )
    {
      auto block = false;
      std::size_t indent = 1;
      std::size_t numindents = 0;

      auto result = std::string{};
      result.reserve( data.size() );
      auto parts = std::vector<std::string>{};
      parts.reserve( 8 );
      boost::algorithm::split( parts, data, boost::is_any_of( "\r\n" ) );
      auto first = true;

      auto asterix = std::regex( "^\\s*\\* " );
      auto dash = std::regex( "^\\s*- " );
      for ( const auto& line : parts )
      {
        if ( !first ) result.append( "\n" );
        if ( std::regex_search( line, asterix ) )
        {
          auto idx = line.find( '*' );
          if ( !block )
          {
            result.append( R"(\begin{itemize}
)"sv );
            block = true;
          }
          else if ( idx > indent )
          {
            result.append( R"(\begin{itemize}
)"sv );
            ++numindents;
          }
          else if ( idx < indent )
          {
            result.append( R"(\end{itemize}
)"sv );
            --numindents;
          }

          indent = idx;
          result.append( R"(\item )" ).append( line.substr( idx + 2 ) );
        }
        else if ( std::regex_search( line, dash ) )
        {
          auto idx = line.find( '-' );
          if ( !block )
          {
            result.append( R"(\begin{itemize}
)"sv );
            block = true;
          }
          result.append( R"(\item )"sv ).append( line.substr( idx + 2 ) );
        }
        else if ( block && line.empty() )
        {
          block = false;
          result.append( R"(\end{itemize}
)"sv );
          while ( numindents > 0 )
          {
            result.append( R"(\end{itemize}
)"sv );
            --numindents;
          }
        }
        else result.append( line );

        first = false;
      }

      return result;
    }

    std::string link( const std::string& data )
    {
      auto bs = data.find( '[' );
      if ( bs == std::string::npos ) return data;

      auto result = std::string{};
      result.reserve( data.size() );
      std::size_t pos = 0;

      while ( bs != std::string::npos )
      {
        result.append( data.substr( pos, bs - pos ) );

        auto be = data.find( "](", bs );
        if ( be == std::string::npos ) return data;
        auto e = data.find( ')', be );
        if ( e == std::string::npos ) return data;

        result.append( R"(\href{)" ).
          append( data.substr( be + 2, e - be - 2 ) ).append( "}{" ).
          append( data.substr( bs + 1, be - bs - 1 ) ).append( "}" );

        pos = e + 1;
        bs = data.find( '[', e );
      }

      if ( pos < result.size() ) result.append( data.substr( pos ) );

      return result;
    }

    std::string style( std::string_view data, std::string_view delimiter, std::string_view cmd )
    {
      auto idx = data.find( delimiter );
      if ( idx == std::string_view::npos ) return std::string{ data };

      auto result = std::string{};
      result.reserve( data.size() );

      auto start = idx + delimiter.size();
      auto end = data.find( delimiter, start + 1 );
      if ( end == std::string_view::npos ) return std::string{ data };

      if ( idx > 0 ) result.append( data.substr( 0, idx ) );
      result.append( cmd ).append( "{" ).
        append( data.substr( start, end - start ) ).append( "}" );

      if ( const auto e = end + delimiter.size(); e < data.size() ) result.append( data.substr( e ) );

      return style( result, delimiter, cmd );
    }

    std::string style( std::string_view data )
    {
      auto result = std::string{};
      result.reserve( data.size() );
      auto parts = std::vector<std::string>{};
      parts.reserve( 8 );
      boost::algorithm::split( parts, data, boost::is_any_of( "\r\n" ) );
      auto first = true;

      for ( const auto& line : parts )
      {
        if ( !first ) result.append( "\n" );
        if ( line.empty() ) continue;

        auto processed = style( line, "`"sv, R"(\texttt)"sv );
        processed = style( processed, "**"sv, R"(\textbf)"sv );
        processed = style( processed, "__"sv, R"(\textbf)"sv );
        processed = style( processed, "*"sv, R"(\textit)"sv );
        processed = style( processed, "_"sv, R"(\textit)"sv );
        boost::algorithm::replace_all( processed, "_", R"(\textunderscore )" );

        result.append( processed );
        first = false;
      }

      return result;
    }
  }
}

std::string spt::output::convert( std::string_view data )
{
  auto contents = std::string{ data };
  boost::algorithm::replace_all( contents, "$", R"(\$)" );
  contents = pconvert::heading( contents );
  contents = pconvert::listing( contents );
  contents = pconvert::bullets( contents );
  contents = pconvert::link( contents );
  contents = pconvert::style( contents );
  boost::algorithm::replace_all( contents, "#", R"(\#)" );
  boost::algorithm::replace_all( contents, "%", R"(\%)" );
  return contents;
}